lib/pid/src/PID.cpp

```diff
@@ -1,492 +1,123 @@
-  pParam_ = Kc;
-  iParam_ = tauI;
-  dParam_ = tauD;
-
-  float tempTauR;
-
-  if (tauI == 0.0) {
-    tempTauR = 0.0;
-  } else {
-    tempTauR = (1.0 / tauI) * tSample_.count();
-  }
-
-  // For "bumpless transfer" we need to rescale the accumulated error.
-  if (inAuto) {
-    if (tempTauR == 0.0) {
-      accError_ = 0.0;
-    } else {
-      accError_ *= (Kc_ * tauR_) / (Kc * tempTauR);
-    }
-  }
-
-  Kc_   = Kc;
-  tauR_ = tempTauR;
-  tauD_ = tauD / tSample_.count();
+void PID::Pid::updateDeadzone(float deadzone) {
+  std::lock_guard<Mutex> lock(m_mutex);
+  m_deadzone = deadzone;
 }
 
-void PID::reset(void) {
-  float scaledBias = 0.0;
-
-  if (usingFeedForward) {
-    scaledBias = (bias_ - outMin_) / outSpan_;
-  } else {
-    scaledBias = (realOutput_ - outMin_) / outSpan_;
-  }
-
-  prevControllerOutput_ = scaledBias;
-  prevProcessVariable_  = (processVariable_ - inMin_) / inSpan_;
-
-  // Clear any error in the integral.
-  accError_ = 0;
-}
-
-void PID::setMode(int mode) {
-  // We were in manual, and we just got set to auto.
-  // Reset the controller internals.
-  if (mode != 0 && !inAuto) {
-    reset();
-  }
-
-  inAuto = (mode != 0);
+uint32_t PID::Pid::reportProportionalGain() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_PGain;
 }
 
-void PID::setInterval(std::chrono::duration<float> interval) {
-  if (interval > 0s) {
-    // Convert the time-based tunings to reflect this change.
-    tauR_ *= (interval / tSample_);
-    accError_ *= (tSample_ / interval);
-    tauD_ *= (interval / tSample_);
-    tSample_ = interval;
-  }
-}
-
-void PID::setSetPoint(float sp) {
-  setPoint_ = sp;
+uint32_t PID::Pid::reportIntegralGain() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_IGain;
 }
 
-void PID::setProcessValue(float pv) {
-  processVariable_ = pv;
+uint32_t PID::Pid::reportDerivativeGain() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_DGain;
 }
 
-void PID::setBias(float bias) {
-  bias_            = bias;
-  usingFeedForward = 1;
+float PID::Pid::reportDeadzone() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_deadzone;
 }
 
-void PID::setDeadZoneError(float error) {
-  deadZoneError_ = error;
+void PID::Pid::reset() {
+  std::lock_guard<Mutex> lock(m_mutex);
+  m_IPath     = 0;
+  m_pastError = 0;
+  m_pastPV    = 0;
+  m_timer.stop();
+  m_timer.reset();
 }
 
-void PID::setRealOutput(float realOutput) {
-  realOutput_ = realOutput;
+float PID::Pid::computePPath(float error) {
+  // no mutex lock needed since inside compute() only
+  return error * m_PGain;
 }
 
-void PID::setupAutoTune(void *outputPointer, float *inputPointer, int actuatorType) {
-  // set actuator type
-  actuatorType_ = actuatorType;
-
-  // set I/O pointers
-  output_ = outputPointer;
-  input_  = inputPointer;
+float PID::Pid::computeIPath(float error, int64_t dt) {
+  // no mutex lock needed since inside compute() only
+  m_IPath += error * dt * m_IGain;
```

Comment:
may be better to use a local var here and update the accumulated i value inside compute. that way all the member var updates are in the same place.

Reply:
hmm then we cant call this function computeIPath. do u have a good name the function should have?

Reply:
tbh i don't like calling them paths at all. i've never heard someone refer to like the contribution of the individual terms as "paths" but that might just be me

Reply:
The matlab series on PID calls em paths: https://www.youtube.com/watch?v=NVLXCwc8HzM&t=571s&ab_channel=MATLAB 3:50